# Overview

eggnog-mapper is a tool for fast functional annotation of novel sequences (genes or proteins) using precomputed eggNOG-based orthology assignments. Obvious examples include the annotation of novel genomes, transcriptomes or even metagenomic gene catalogs. The use of orthology predictions for functional annotation is considered more precise than traditional homology searches, as it avoids transferring annotations from paralogs (duplicate genes with a higher chance of being involved in functional divergence).

EggNOG-mapper is also available as a public online resource: http://beta-eggnogdb.embl.de/#/app/emapper.

# Documentation
https://github.com/jhcepas/eggnog-mapper/wiki

# Citation

If you use this software, please cite:
```

[1] Fast genome-wide functional annotation through orthology assignment by
     eggNOG-mapper. Jaime Huerta-Cepas, Kristoffer Forslund, Luis Pedro Coelho,
     Damian Szklarczyk, Lars Juhl Jensen, Christian von Mering and Peer Bork.
     Mol Biol Evol (2017). [doi:
     10.1093/molbev/msx148](https://doi.org/10.1093/molbev/msx148)

[2] eggNOG 4.5: a hierarchical orthology framework with improved functional
      annotations for eukaryotic, prokaryotic and viral sequences. Jaime
      Huerta-Cepas, Damian Szklarczyk, Kristoffer Forslund, Helen Cook, Davide
      Heller, Mathias C. Walter, Thomas Rattei, Daniel R. Mende, Shinichi
      Sunagawa, Michael Kuhn, Lars Juhl Jensen, Christian von Mering, and Peer
      Bork. Nucl. Acids Res. (04 January 2016) 44 (D1): D286-D293. [doi:
      10.1093/nar/gkv1248](http://doi.org/10.1093/nar/gkv1248)
```
